vllm/model_executor/layers/sampler.py

```diff
@@ -59,25 +69,50 @@ def forward(
         assert len(presence_penalties) == logits.shape[0]
         assert len(frequency_penalties) == logits.shape[0]
         logits = _apply_penalties(logits, output_tokens, presence_penalties,
-                                  frequency_penalties, self.vocab_size)
-
-        # Apply temperature scaling.
-        temperatures = _get_temperatures(input_metadata)
-        assert len(temperatures) == logits.shape[0]
-        if any(t != 1.0 for t in temperatures):
-            t = torch.tensor(temperatures,
-                             dtype=logits.dtype,
-                             device=logits.device)
-            # Use in-place division to avoid creating a new tensor.
-            logits.div_(t.unsqueeze(dim=1))
-
-        # Apply top-p and top-k truncation.
-        top_ps, top_ks = _get_top_p_top_k(input_metadata, self.vocab_size)
-        assert len(top_ps) == len(top_ks) == logits.shape[0]
-        do_top_p = any(p < 1.0 - _SAMPLING_EPS for p in top_ps)
-        do_top_k = any(k != self.vocab_size for k in top_ks)
-        if do_top_p or do_top_k:
-            logits = _apply_top_p_top_k(logits, top_ps, top_ks)
+                                  frequency_penalties)
+
+        # Sort logits by sampling type.
+        sampling_type_indices = input_metadata.sampling_type_indices
+        sampling_type_indices_values, sampling_type_indices_index = (
+            sampling_type_indices.sort())
+        # We use the bincount+cumsum to compute the offsets.
+        sampling_type_offsets = torch.bincount(sampling_type_indices_values,
+                                               minlength=3).cumsum(0)[:-1]
+        # The logit_to_id_map lets us reverse the sort we did above when we
+        # match outputs to sequences.
+        logit_to_id_map = sampling_type_indices_index.argsort()
+        logits = logits[sampling_type_indices_index]
+
+        # If there are only greedy requests (and therefore, no requests
+        # we can apply temperature & top_p/top_k to), we can return early.
+        # Otherwise, we will slice the logits and modify only the non-greedy
+        # subset.
+        if sampling_type_indices.max() > SamplingType.GREEDY:
+            non_greedy_offset = sampling_type_offsets[0]
```

Comment:
nit: consider 
`sampling_type_offsets[SamplingType.GREEDY]`
and other places too.